File: entroly-wasm/test_work_graph_repo.js
'use strict';

const crypto = require('crypto');
const fs = require('fs');
const os = require('os');
const path = require('path');
const zlib = require('zlib');
const { spawnSync } = require('child_process');
const {
  RepositoryDiscoveryError,
  discoverRepositoryIdentity,
  discoverRepositoryObservation,
} = require('./js/work_graph_repo');

function assert(condition, message) {
  if (!condition) throw new Error(message || 'assertion failed');
}

function git(repo, ...args) {
  const result = spawnSync('git', ['-C', repo, ...args], { encoding: 'utf8' });
  if (result.status !== 0) throw new Error(result.stderr || `git ${args.join(' ')} failed`);
  return String(result.stdout || '').trim();
}

const root = fs.mkdtempSync(path.join(os.tmpdir(), 'entroly-workgraph-'));
const repo = path.join(root, 'repo');
fs.mkdirSync(repo);

try {
  git(repo, 'init', '-b', 'main');
  git(repo, 'config', 'user.email', 'test@example.com');
  git(repo, 'config', 'user.name', 'Test');
  fs.writeFileSync(path.join(repo, 'app.js'), 'console.log("one")\n');
  git(repo, 'add', 'app.js');
  git(repo, 'commit', '-m', 'initial');

  const clean = discoverRepositoryObservation(repo, { observedAtMs: 1000 });
  assert(clean.branch.name === 'main', 'clean branch mismatch');
  assert(clean.branch.ahead_by === 0, 'clean repo must not be ahead');
  assert(clean.changes.length === 0, 'clean repo must have no worktree changes');
  assert(clean.commits.length === 0, 'clean repo must remain a null-control observation');
  assert(clean.task_hint === null, 'observer must not invent task intent');

  const identity = discoverRepositoryIdentity(repo);
  assert(identity.repo_id === clean.repo_id, 'identity-only repo ID drifted from full observation');
  assert(path.resolve(identity.root) === path.resolve(repo), 'identity root mismatch');

  const explicit = discoverRepositoryObservation(repo, {
    defaultBranch: 'main', observedAtMs: 1001,
  });
  assert(explicit.branch.default_branch === 'main', 'valid branch override rejected');
  assert(explicit.branch.base_ref === 'refs/heads/main', 'explicit base ref mismatch');
  let invalidRejected = false;
  try { discoverRepositoryObservation(repo, { defaultBranch: '--help' }); }
  catch (error) { invalidRejected = error instanceof RepositoryDiscoveryError; }
  assert(invalidRejected, 'unsafe default branch override was accepted');

  const checkpointDir = path.join(root, 'checkpoints');
  fs.mkdirSync(checkpointDir);
  const checkpoint = {
    checkpoint_id: 'ckpt_test_1',
    metadata: {
      task: 'Fix streaming',
      step: 'finish tests',
      decisions: ['Preserve event order'],
    },
  };
  fs.writeFileSync(
    path.join(checkpointDir, 'ckpt_test_1.json.gz'),
    zlib.gzipSync(Buffer.from(JSON.stringify(checkpoint))),
  );
  const cleanWithCheckpoint = discoverRepositoryObservation(repo, {
    checkpointDir, observedAtMs: 1100,
  });
  assert(cleanWithCheckpoint.task_hint === null, 'stale checkpoint resurrected work in clean repo');
  assert(cleanWithCheckpoint.decisions.length === 0, 'stale decisions leaked into clean repo');

  git(repo, 'checkout', '-b', 'feature/work');
  fs.writeFileSync(path.join(repo, 'app.js'), 'console.log("two")\n');
  git(repo, 'add', 'app.js');
  git(repo, 'commit', '-m', 'feature change');
  fs.writeFileSync(path.join(repo, 'app.js'), 'console.log("three")\n');
  fs.writeFileSync(path.join(repo, 'new.txt'), 'new\n');

  const active = discoverRepositoryObservation(repo, {
    agentId: 'codex',
    sessionId: 's1',
    checkpointDir,
    observedAtMs: 2000,
  });
  assert(active.branch.base_ref === 'refs/heads/main', 'base ref mismatch');
  assert(active.branch.ahead_by === 1, `expected one ahead commit, got ${active.branch.ahead_by}`);
  assert(active.commits.length === 1, 'expected one branch commit');
  assert(active.commits[0].subject === 'feature change', 'commit subject mismatch');
  const kinds = Object.fromEntries(active.changes.map(change => [change.path, change.kind]));
  assert(kinds['app.js'] === 'modified', 'modified file missing');
  assert(kinds['new.txt'] === 'untracked', 'untracked file missing');
  assert(active.task_hint.title === 'Fix streaming', 'checkpoint did not annotate real Git work');
  assert(active.task_hint.remaining_work[0] === 'finish tests', 'checkpoint remaining work missing');
  assert(active.decisions[0].text === 'Preserve event order', 'checkpoint decision missing');

  const oldHome = process.env.HOME;
  const oldUserProfile = process.env.USERPROFILE;
  const oldEntrolyDir = process.env.ENTROLY_DIR;
  const fakeHome = path.join(root, 'home');
  fs.mkdirSync(fakeHome);
  process.env.HOME = fakeHome;
  // Node's os.homedir() reads USERPROFILE on Windows and HOME on Unix. Set
  // both so this test controls the same API on every supported platform.
  process.env.USERPROFILE = fakeHome;
  delete process.env.ENTROLY_DIR;
  try {
    const absentRoot = path.join(fakeHome, '.entroly');
    const beforeDefault = discoverRepositoryObservation(repo, {
      includeCheckpoint: true,
      observedAtMs: 2100,
    });
    assert(beforeDefault.task_hint === null, 'missing default checkpoint invented a task');
    assert(!fs.existsSync(absentRoot), 'checkpoint lookup created ~/.entroly as a read side effect');

    const projectHash = crypto.createHash('sha256')
      .update(path.resolve(repo), 'utf8').digest('hex').slice(0, 12);
    const defaultDir = path.join(fakeHome, '.entroly', 'checkpoints', projectHash);
    fs.mkdirSync(defaultDir, { recursive: true });
    fs.copyFileSync(
      path.join(checkpointDir, 'ckpt_test_1.json.gz'),
      path.join(defaultDir, 'ckpt_test_1.json.gz'),
    );
    const defaultCheckpoint = discoverRepositoryObservation(repo, {
      includeCheckpoint: true,
      observedAtMs: 2200,
    });
    assert(defaultCheckpoint.task_hint.title === 'Fix streaming', 'default checkpoint path parity failed');
  } finally {
    if (oldHome === undefined) delete process.env.HOME; else process.env.HOME = oldHome;
    if (oldUserProfile === undefined) delete process.env.USERPROFILE; else process.env.USERPROFILE = oldUserProfile;
    if (oldEntrolyDir === undefined) delete process.env.ENTROLY_DIR; else process.env.ENTROLY_DIR = oldEntrolyDir;
  }

  git(repo, 'remote', 'add', 'origin', 'https://alice:secret-one@example.com/org/repo.git');
  const firstId = discoverRepositoryObservation(repo, { observedAtMs: 3000 }).repo_id;
  git(repo, 'remote', 'set-url', 'origin', 'https://bob:secret-two@example.com/org/repo.git');
  const secondId = discoverRepositoryObservation(repo, { observedAtMs: 3001 }).repo_id;
  assert(firstId === secondId, 'credential changes must not change repo identity');
  assert(!firstId.includes('secret') && !firstId.includes('alice'), 'repo identity leaked credentials');

  const hugeRepo = path.join(root, 'huge');
  fs.mkdirSync(hugeRepo);
  git(hugeRepo, 'init', '-b', 'main');
  git(hugeRepo, 'config', 'user.email', 'test@example.com');
  git(hugeRepo, 'config', 'user.name', 'Test');
  fs.writeFileSync(path.join(hugeRepo, 'base.txt'), 'base\n');
  git(hugeRepo, 'add', 'base.txt');
  git(hugeRepo, 'commit', '-m', 'initial');
  for (let index = 0; index < 513; index += 1) {
    fs.writeFileSync(path.join(hugeRepo, `untracked-${index}.txt`), 'x\n');
  }
  const hugeIdentity = discoverRepositoryIdentity(hugeRepo);
  assert(hugeIdentity.repo_id.startsWith('git-root:'), 'identity-only lookup failed on huge dirty repo');
  const hugeObservation = discoverRepositoryObservation(hugeRepo);
  assert(hugeObservation.changes.length === 513,
    'large dirty observation silently truncated changed paths');

  console.log('Work Graph npm Git discovery contract: PASS');
} finally {
  fs.rmSync(root, { recursive: true, force: true });
}

File: tests/test_work_graph_repo.py
from __future__ import annotations

import gzip
import json
import subprocess
from pathlib import Path

import pytest

from entroly.work_graph_repo import RepositoryDiscoveryError, discover_repository_observation


def _git(repo: Path, *args: str) -> str:
    result = subprocess.run(
        ["git", "-C", str(repo), *args],
        check=True,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    return result.stdout.strip()


def _repo(tmp_path: Path) -> Path:
    repo = tmp_path / "repo"
    repo.mkdir()
    _git(repo, "init", "-b", "main")
    _git(repo, "config", "user.email", "test@example.com")
    _git(repo, "config", "user.name", "Test")
    (repo / "app.py").write_text("print('one')\n")
    _git(repo, "add", "app.py")
    _git(repo, "commit", "-m", "initial")
    return repo


def test_clean_repo_is_null_control(tmp_path: Path) -> None:
    repo = _repo(tmp_path)
    observation = discover_repository_observation(repo, observed_at_ms=1234)

    assert observation["observed_at_ms"] == 1234
    assert observation["branch"]["name"] == "main"
    assert observation["branch"]["default_branch"] == "main"
    assert observation["branch"]["ahead_by"] == 0
    assert observation["branch"]["behind_by"] == 0
    assert observation["changes"] == []
    assert observation["commits"] == []
    assert observation["task_hint"] is None
    assert observation["decisions"] == []


def test_feature_branch_observes_only_durable_work_facts(tmp_path: Path) -> None:
    repo = _repo(tmp_path)
    _git(repo, "checkout", "-b", "feature/stream")
    (repo / "app.py").write_text("print('two')\n")
    _git(repo, "add", "app.py")
    _git(repo, "commit", "-m", "change app")
    (repo / "app.py").write_text("print('three')\n")
    (repo / "new.txt").write_text("new\n")

    observation = discover_repository_observation(
        repo,
        agent_id="claude",
        session_id="session-1",
        task_hint={
            "task_id": "stream",
            "title": "Fix streaming",
            "trust": "observed",
            "source_kind": "user_statement",
            "source_ref": "user:task",
        },
        observed_at_ms=2000,
    )

    assert observation["repo_id"].startswith("git-root:")
    assert observation["agent_id"] == "claude"
    assert observation["session_id"] == "session-1"
    assert observation["task_hint"]["title"] == "Fix streaming"
    assert observation["branch"]["name"] == "feature/stream"
    assert observation["branch"]["base_ref"] == "refs/heads/main"
    assert observation["branch"]["ahead_by"] == 1
    assert len(observation["commits"]) == 1
    assert observation["commits"][0]["subject"] == "change app"
    kinds = {item["path"]: item["kind"] for item in observation["changes"]}
    assert kinds == {"app.py": "modified", "new.txt": "untracked"}


def test_rename_and_staged_state_are_preserved(tmp_path: Path) -> None:
    repo = _repo(tmp_path)
    _git(repo, "mv", "app.py", "renamed.py")
    observation = discover_repository_observation(repo, observed_at_ms=3000)

    assert len(observation["changes"]) == 1
    change = observation["changes"][0]
    assert change["kind"] == "renamed"
    assert change["path"] == "renamed.py"
    assert change["old_path"] == "app.py"
    assert change["staged"] is True


def test_remote_credentials_never_affect_or_leak_repo_identity(tmp_path: Path) -> None:
    repo = _repo(tmp_path)
    _git(repo, "remote", "add", "origin", "https://alice:secret-one@example.com/org/repo.git")
    first = discover_repository_observation(repo, observed_at_ms=4000)["repo_id"]
    _git(repo, "remote", "set-url", "origin", "https://bob:secret-two@example.com/org/repo.git")
    second = discover_repository_observation(repo, observed_at_ms=4001)["repo_id"]

    assert first == second
    assert "secret" not in first
    assert "alice" not in first
    assert "bob" not in second


def test_detached_head_is_observed_without_inventing_task(tmp_path: Path) -> None:
    repo = _repo(tmp_path)
    sha = _git(repo, "rev-parse", "HEAD")
    _git(repo, "checkout", "--detach", sha)
    observation = discover_repository_observation(repo, observed_at_ms=5000)

    assert observation["branch"]["detached"] is True
    assert observation["branch"]["name"] == ""
    assert observation["task_hint"] is None


def test_invalid_default_branch_override_is_rejected(tmp_path: Path) -> None:
    repo = _repo(tmp_path)
    with pytest.raises(RepositoryDiscoveryError, match="invalid default branch"):
        discover_repository_observation(repo, default_branch="--help")


def test_valid_default_branch_override_is_accepted(tmp_path: Path) -> None:
    repo = _repo(tmp_path)
    observation = discover_repository_observation(repo, default_branch="main")
    assert observation["branch"]["default_branch"] == "main"
    assert observation["branch"]["base_ref"] == "refs/heads/main"


def test_large_dirty_repo_is_complete_instead_of_silently_truncated(tmp_path: Path) -> None:
    repo = _repo(tmp_path)
    for index in range(513):
        (repo / f"untracked-{index}.txt").write_text("x\n")
    observation = discover_repository_observation(repo)
    assert len(observation["changes"]) == 513
    paths = {change["path"] for change in observation["changes"]}
    assert {"untracked-0.txt", "untracked-512.txt"} <= paths


def test_checkpoint_can_name_existing_git_work_but_not_resurrect_clean_repo(tmp_path: Path) -> None:
    repo = _repo(tmp_path)
    ckpt_dir = tmp_path / "checkpoints"
    ckpt_dir.mkdir()
    payload = {
        "schema_version": 2,
        "checkpoint_id": "ckpt_test_1000_1",
        "timestamp": 1000.0,
        "current_turn": 1,
        "fragments": [],
        "dedup_fingerprints": {},
        "co_access_data": {},
        "metadata": {
            "task": "Fix streaming",
            "step": "finish tests",
            "decisions": ["Preserve event order"],
        },
        "stats": {},
    }
    with gzip.open(ckpt_dir / "ckpt_test_1000_1.json.gz", "wt", encoding="utf-8") as handle:
        json.dump(payload, handle)

    clean = discover_repository_observation(repo, checkpoint_dir=ckpt_dir)
    assert clean["task_hint"] is None
    assert clean["decisions"] == []

    (repo / "app.py").write_text("print('working')\n")
    dirty = discover_repository_observation(repo, checkpoint_dir=ckpt_dir)
    assert dirty["task_hint"]["title"] == "Fix streaming"
    assert dirty["task_hint"]["source_kind"] == "checkpoint"
    assert dirty["task_hint"]["remaining_work"] == ["finish tests"]
    assert [item["text"] for item in dirty["decisions"]] == ["Preserve event order"]
